openhcl/underhill_mem/src/lib.rs

```diff
@@ -488,35 +486,38 @@ impl ProtectIsolatedMemory for HardwareIsolatedMemoryProtector {
         // vtl permissions to the page.
         let orig_gpns = gpns;
         let mut failed_vtl_permission_index = None;
-        let gpns = gpns
-            .iter()
-            .copied()
-            .enumerate()
-            .take_while(|&(index, gpn)| {
-                if vtl == GuestVtl::Vtl0 && shared && self.vtl1_protections_enabled() {
-                    let permissions = self
-                        .vtl0
-                        .query_access_permission(gpn)
-                        .expect("vtl 1 protections enabled, vtl permissions should be tracked");
-                    if !permissions.readable() || !permissions.writable() {
-                        // TODO GUEST VSM: should this send a memory intercept instead?
-                        failed_vtl_permission_index = Some(index);
-                        false
+        let gpns = {
+            gpns.iter()
+                .copied()
+                .enumerate()
+                .take_while(|&(index, gpn)| {
+                    if vtl == GuestVtl::Vtl0 && shared && self.vtl1_protections_enabled() {
+                        let op = || {
+                            let permissions = self.vtl0.query_access_permission(gpn).expect(
+                                "vtl 1 protections enabled, vtl permissions should be tracked",
+                            );
+                            if !permissions.readable() || !permissions.writable() {
+                                failed_vtl_permission_index = Some(index);
+                                false
+                            } else {
+                                true
+                            }
+                        };
+
+                        guestmem::rcu().run(op)
```

Comment:
I think it's more idiomatic to just do:

```
guestmem::rcu().run(|| {
    ...
})
```

rather than bind the closure to variable.

Reply:
Also, if I'm not mistaken, it seems like mutations to this are already protected by the lock we're holding. So I don't think you need to use an RCU critical section here.